### PR TITLE
Attempt to add some info about design principles

### PR DESCRIPTION
**Pull request by yuvipanda**
This isn't complete, and possibly subjective. Lots of review is welcome &
required.

I've attempted to put down some of the design principles in use in this repo,
and codify information about some of the design decisions I've made about
this code. Writing this down somewhere (maybe README.md is not the best
location?) is important for the long term health of this codebase. It's a somewhat
complex codebase with wide ranging effects in many other projects (such as
BinderHub), so this is quite important to get right.

### CODE REVIEW

**Comment by betatim** [comment]
Maybe move to `docs/design.md` or so?

Otherwise LGTM.

File: README.md [hunk 1]
```diff
@@ -100,3 +100,74 @@ be as they are released.
...
+composable, and need to be able to work well with other languages.
+
+### [Pareto principle](https://en.wikipedia.org/wiki/Pareto_principle)
+
+Roughly speaking, we want to support 80% of use cases, and provide an escape
+hatch (raw Dockerfiles) for the other 20%. We explicitly want to provide support
+only for the most common use cases - covering every possible use case never ends
+well.
+
+An easy process for getting support for more languages here is to demonstrate
```
Comment (willingc) [inline README.md]: Replace last paragraph:

For a pattern (a language or combination of languages) to be included inside the `repo2docker` repository, the pattern and its associated Dockerfile must have value to many other people beyond the creator.

File: README.md [hunk 2]
```diff
@@ -100,3 +100,74 @@ be as they are released.
...
+   we know it is going to be the same. This provides massive performance &
+   architectural advantages when building additional tools (like BinderHub) on
+   top of repo2docker.
+2. We produce Dockerfiles that have as much in common as possible across
+   multiple repos, enabling better use of the Docker build cache. This also
+   provides massive performance advantages.
+
+### Unix principles
+
+`repo2docker` should do one thing, and do it well. This one thing is:
```
Comment (choldgraf) [inline README.md]: I think this sentence or something close to it should be higher up in the readme, and on the `index` page of the docs

File: README.md [hunk 3]
```diff
@@ -100,3 +100,74 @@ be as they are released.
...
+
+> Given a repository (of some sort), deterministically build a docker image from
+> it.
+
+There's also some convenience code (to run the built image) for users, but
```
[the remark this answers is not in the document]
Comment (choldgraf) [reply README.md]: agree - it also scopes what we expect the package to do more cleanly, e.g., we aren't going to try and build fancier running functionality into the project

File: README.md [hunk 4]
```diff
@@ -100,3 +100,74 @@ be as they are released.
...
+
+### Composability
+
+The prime reason `repo2docker` exists (rather than just using something
+like [s2i](https://github.com/openshift/source-to-image)) is we want to support
+*composable* environments. We want to easily have an image with
+Python3+Julia+R-3.2 environments, rather than just one single language
+environment. While generally one language environment per container works well,
+in many scientific / datascience computing environments you need multiple
+languages working together to get anything done. So all buildpacks are
```
Comment (choldgraf) [inline README.md]: sounds a bit harsh...maybe `...languages working together to get your work done.`

File: README.md [hunk 5]
```diff
@@ -100,3 +100,74 @@ be as they are released.
...
+composable, and need to be able to work well with other languages.
+
+### [Pareto principle](https://en.wikipedia.org/wiki/Pareto_principle)
+
+Roughly speaking, we want to support 80% of use cases, and provide an escape
+hatch (raw Dockerfiles) for the other 20%. We explicitly want to provide support
+only for the most common use cases - covering every possible use case never ends
```
[the remark this answers is not in the document]
Comment (choldgraf) [reply README.md]: I like to think of this as an extreme pareto principle (95 / 5) but I get the point :-)

**Comment by yuvipanda** [comment]
Thanks to @Carreau for doing most of the hard work :)

I've:

1. Moved it to docs/source
2. Added it to the toc

**Comment by Carreau** [comment]
@willingc is supposed to be in holidays, so I'm going to merge for her, if all goes fine she gets the credits, if we have complain I'll take the blame. 

Thanks !

**Pull request merged**